### PR TITLE
Config Singleton

### LINKED ISSUE #1158: Config Singleton
It would be really nice if libvalhalla had a config singleton. Something where we initialize it once and then can access it globally after that. Why would we want that...

Right now we have the annoying problem of having to pass a config to all of the top level elements. It would be nice at program start to just configure the singleton and let these objects get access to the config on the fly.

One problem this would solve directly would be that we have config values that get validated against for requests. Currently a loki worker will load these when its instantiated and then validate each request against them. This makes it so that some of the things we'd like to pull out to shared request object cannot be because the code that genrates this pbf object doesnt have access to the config only loki_worker.

Basically having a globally configured valhalla is just convenient :smile: 

more details:

1. singleton getter takes an optional config file so that on first call it gets configured (i think the python bindings used to be like this, we'd just move this idea into the library)
2. the config object knows whether or not its configured. if its not configured we throw in the getter or maybe even assert to force kill the application? this forces who ever is using it to make sure their use includes configuring the library first.
3. wrap the config boilerplate that every excecutable will have into the arg parsing boilerplate that we are adding in #4169

### PR DESCRIPTION
**Pull request by janusz-anue**
Fixes #1158 

Sorry for the long wait. I have finally gotten around tackling this PR. As discussed in #4177 and in #1158, this PR solves the first phase of implementing the configuration as a singleton and changes the top-level passings. This means that basically the arg_parse is changed and every class using it, as well as the valhalla_service.
Later in another phase the other passings can be removed too.

Before using config() the singleton should be configured with a config file or an inline config string.

I am open for suggestions and also nitpicks as I am looking to improve my c++ coding.

Janusz Spatz on behalf of Mercedes-Benz Tech Innovation GmbH.
[Provider Information](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md)

### CODE REVIEW

**Comment by nilsnolde** [comment]
Thanks a lot @janusz-anue ! We'll take a look in the next days.

FYI, ever since your last commit one more exe has been added in `mjolnir/valhalla_build_landmarks.cc` which fails the build currently.

File: valhalla/configuration.h
Comment (nilsnolde) [inline valhalla/configuration.h:38]: IMO that's a bit convoluted here. I also might not fully be aware of all constraints, but I'd prefer if this would immediately return the `config_`. I think you could remove the 2 free functions then, and the first call to this would need to specify a valid path to a valhalla.json or it'll throw in the constructor as it currently does. I think it's basically the same as your concept, just with a bit less code/fewer functions and all encapsulated in the class. Also, free functions with such generic names could be problematic at some point. You could call the function maybe rather `get_config` or so, it wouldn't be an instance anymore at that point. The important thing is that both function & `Configuration` are static. We could also return a `Configuration` instance from another function if we'd want to add more useful functions to `Configuration` class, which might be interesting.

Does that make sense?

**Review by nilsnolde — COMMENTED** [review]
Thanks, looks good @janusz-anue. One point for discussion in the singleton design. WDYT?

**Comment by nilsnolde** [comment]
You'll also need to format the code, see https://github.com/valhalla/valhalla/blob/master/CONTRIBUTING.md#code-contributions for pre-commit which makes it a bit easier IMO.

File: valhalla/configuration.h
Comment (nilsnolde) [reply valhalla/configuration.h:38]: Another nit might be to call the class `config_singleton_t` instead of `Configuration`, following the existing pattern around singletons.

Comment (janusz-anue) [reply valhalla/configuration.h:38]: I think I have an idea what you mean, I think it makes sense from an encapsulation point of view, will update.

Comment (nilsnolde) [reply valhalla/configuration.h:38]: FYI, was just following up with @kevinkresier last week and he’d prefer to have a free function in the valhalla namespace, behaving similarly as I described above, so that we can just do valhalla::config(<optional_path>).

Comment (janusz-anue) [reply valhalla/configuration.h]: I have adjusted the code a bit and taken out the configure-function, is it like you imagined?

Comment (nilsnolde) [reply valhalla/configuration.h]: Jep, thanks a lot!

**Review by nilsnolde — DISMISSED** [review]
LGTM! Only thing left is a CHANGELOG entry.

File: valhalla/configuration.h
Comment (kevinkreiser) [inline valhalla/configuration.h:38]: just move this down into the namespace level function below and remove this function. we dont need two ways to do the same thing, do we?

File: valhalla/configuration.h
Comment (kevinkreiser) [inline valhalla/configuration.h:10]: since we are on c++17 now you can use std::filesystem instead. we should try to get away from using the our port over time.

File: valhalla/configuration.h
Comment (kevinkreiser) [reply valhalla/configuration.h:38]: (you'd need to add the namespace function as a friend function of this class to make that work i guess)

File: valhalla/configuration.h
Comment (janusz-anue) [reply valhalla/configuration.h]: Yeah, after doing that I needed to resolve some naming collisions (config function and config variable in anonymous namespace) in some testifies. I resolved them by just renaming the var to cfg. Hope that's ok

File: valhalla/configuration.h
Comment (janusz-anue) [reply valhalla/configuration.h]: Added the function as a friend, needed to declare the function before defining it, that's why I added the declaration above the singleton

**Comment by janusz-anue** [comment]
When playing around with the new singleton I was noticing that when used with a long inline config the check to see if the argument is a file or inline config threw an exception. This is now mitigated in the new try-catch block in the constructor.

Additionally I was finding that when I read the singleton in the GraphTile for example, many tests will fail because the tests (that use the GraphTile in any way) logically don't initialise the singleton. So I thought to create a new exception for when the singleton is not initialised that can be caught and handled accordingly. I can expand on that in a new (draft) PR. Another solution could be to add a member var that is true when the config is initialised, otherwise false.

**Comment by johannes-no** [comment]
Hi @kevinkreiser and/or @nilsnolde, could you please take another look at this PR so that we can use the config singleton to complete the original task [#4177](https://github.com/valhalla/valhalla/issues/4177) . For the speed fading the PR is prepared but i would wait to open it until the merge of this PR.

File: CHANGELOG.md
Comment (kevinkreiser) [inline CHANGELOG.md:26]: would you mind moving this to the bottom of the list

File: test/isochrone.cc
Comment (kevinkreiser) [inline test/isochrone.cc:33]: yeah this is an inconvenience. not being able to name something `config` when inside the valhalla namespace, the good news is that in a future pr we wont need to even keep the variable around we can just always call `config()` whereever we are in the code (in most scenarios)

File: valhalla/config.h
Comment (kevinkreiser) [inline valhalla/config.h:60]: why not just use `throw std::runtime_error("Config singleton was not initialized before usage");`? it must be because we have try catch blocks in things like the workers which key off the type of exception? perhaps the workers have a new catch clause that returns a specific error when not configured? or maybe it should be a valhalla_exception_t with a specific error code? fwiw this is fine, dont feel the need to change it, i more just wanted to discuss it/think about it a little.

File: CHANGELOG.md
Comment (kevinkreiser) [reply CHANGELOG.md:26]: ill just move this and merge so as not to hold this up any longer. very sorry for the wait

**Review by kevinkreiser — APPROVED** [review]

File: CHANGELOG.md
Comment (johannes-no) [reply CHANGELOG.md]: Thanks Kevin.

**Pull request merged**
